Respect alphabetical order of collections in "Move" modal

kamilmielnik opened the pull request:
Closes #31294
Closes #31425

### Description

- I discovered that there was the same sorting issue with Snippet Collection Picker and got green light from Roman to fix it as well, so the fix is included here.
- I reported this issue I found while working on this: #31394


### How to verify

Currently collections are ordered by id - the earlier collection was created, the earlier it will be on the list in the picker.
We want the collections to be sorted alphabetically instead.
See demos below for exact reproduction steps.

<details>
<summary>
<h3>Demo - Collection Picker</h3>
</summary>

#### Before

https://github.com/metabase/metabase/assets/6830683/b963f134-54ba-48da-8254-1315803c22ed


#### After

https://github.com/metabase/metabase/assets/6830683/9b55fc33-3030-4d22-8bdc-0fab0e7ac362

</details>
<details>
<summary>
<h3>Demo - Snippet Collection Picker</h3>
<p>This can be tested only using the Enterprise Edition.</p>
</summary>

#### Before

https://github.com/metabase/metabase/assets/6830683/891b18d6-ff99-4b04-930c-64f22118a427


#### After

https://github.com/metabase/metabase/assets/6830683/932c4055-28c2-48dd-808d-c238a016e35b


</details>

kamilmielnik committed:
Remove unused openCollection prop

kamilmielnik committed:
Preserve collection sorting given by API in getExpandedCollectionsById
- Fixes #31294

kamilmielnik committed:
Use Collections.getListStatePath() to unhardcode the object path

kamilmielnik committed:
Use CollectionId instead of any

kamilmielnik committed:
Extract Collections.selectors.getCollectionList

kamilmielnik committed:
Convert Collections selectors unit tests to TypeScript

kamilmielnik committed:
Remove redundant selector

kamilmielnik committed:
Add a unit test for Collections.selectorsgetExpandedCollectionsById

kamilmielnik committed:
Use createMockCollection for all test collections
- Use Partial<Collection> instead of MockCollection type

kamilmielnik committed:
Add a test case for non-nested collections

kamilmielnik committed:
Merge branch 'master' into 31294-alphabetical-collections-in-move-modal

kamilmielnik committed:
Rename identifiers

kamilmielnik committed:
Add a test suite for snippet collections

kamilmielnik committed:
Fix SnippetCollections.selectors.getExpandedCollectionsById

kamilmielnik committed:
Refactor tests

kamilmielnik committed:
Refactor tests

kamilmielnik committed:
Merge branch 'master' into 31294-alphabetical-collections-in-move-modal

ranquild reviewed (changes requested):
Added a comment about mocking _list state. Please ping me if you have questions about that

ranquild reviewed (changes requested):
@kamilmielnik As discussed, we don't test reducers (you can search for this in the official redux docs for the rationale). I'd follow the steps from the issue https://github.com/metabase/metabase/issues/31294, render the affected component with mocked BE data, and check that the order of the items is preserved. To simulate the issue you would likely need to add some other existing collections to the `entities` state directly before rendering the component. Please let me know if that works

kamilmielnik committed:
Merge branch 'master' into 31294-alphabetical-collections-in-move-modal

kamilmielnik committed:
Revert unit tests for selectors

kamilmielnik committed:
Revert convert selectors to TypeScript

kamilmielnik committed:
Add error support to setupCollectionsEndpoints

kamilmielnik committed:
Introduce useCollectionsQuery

kamilmielnik committed:
Fix tests

kamilmielnik committed:
Add unit tests for preserving order of collections in ItemPicker

kamilmielnik committed:
Make assertions more semantic

kamilmielnik committed:
Merge branch 'master' into 31294-alphabetical-collections-in-move-modal

kamilmielnik committed:
Merge branch 'master' into 31294-alphabetical-collections-in-move-modal

kamilmielnik committed:
Allow mocking root collection
- Make ROOT_COLLECTION object match Collection type

kamilmielnik committed:
Add tests for preserving order in snippet collections list
- fix case with an undefined snippet collections list

kamilmielnik committed:
Update test case with data loading necessary for ItemPicker underneath the EditSandboxModal after updating selectors

ranquild commented on frontend/src/metabase/common/hooks/use-collections-query/use-collections-query.ts:
```diff
@@ -0,0 +1,20 @@
+  useEntityListQuery,
+} from "../use-entity-list-query";
+
+export const useCollectionsQuery = (
```
`useCollectionListQuery` - please follow the established conventions

kamilmielnik replied:
3b8b52fc54755af96433335406bb470d8132fe94

ranquild commented on frontend/src/metabase/containers/ItemPicker/ItemPicker.unit.spec.js:
```diff
@@ -241,4 +294,134 @@ describe("ItemPicker", () => {
+        COLLECTION.REGULAR_2,
+      ];
+
+      fetchMock.get("path:/api/collection", collections);
```
We setup fetch mocks within the `setup` function, and use mocks defined in `server-mocks`

kamilmielnik replied:
This was a leftover, good catch, removed: 4d444a33f293169f855ace66217f27f6a86c81bf

kamilmielnik replied:
+ 066febb437e5081f5f14a37e007dcc80d2269363

ranquild commented on frontend/test/__support__/server-mocks/collection.ts:
```diff
@@ -14,8 +14,23 @@ import {
 
-export function setupCollectionsEndpoints(collections: Collection[]) {
-  fetchMock.get("path:/api/collection/root", ROOT_COLLECTION);
+export function setupCollectionsEndpoints({
```
We have different functions to setup error cases, and the one for collections already exist - it's `setupUnauthorizedCollectionsEndpoints`. Please use the existing one.

kamilmielnik replied:
I think setup like this would be semantically misleading - after all we're not testing authorization here.
It'd be more appropriate to use a more generic error.

BTW `setupUnauthorizedCollectionsEndpoints` is unused.
I'll reuse it as suggested but first refactor it to allow more generic errors.

kamilmielnik replied:
> We have different functions to setup error cases,

Did just that: eb8e6377729da45557ee926f2c17bda8064563ce

kamilmielnik committed:
Update test case with data loading necessary for ItemPicker underneath the AddToDashSelectDashModal after updating selectors

ranquild reviewed (commented):
@kamilmielnik Looks very good! I've left some comments about using the established patterns, please take a look

kamilmielnik committed:
Rename useCollectionsQuery to useCollectionListQuery

kamilmielnik committed:
Remove leftover fetchMock

kamilmielnik committed:
Move error out of setupCollectionsEndpoints

kamilmielnik committed:
Fix ModelDetailPage unit tests setup

kamilmielnik committed:
Merge branch 'master' into 31294-alphabetical-collections-in-move-modal

kamilmielnik committed:
Remove leftover fetchMocks

ranquild commented on frontend/src/metabase/entities/collections/constants.ts:
```diff
@@ -1,12 +1,17 @@
 export const DEFAULT_COLLECTION_COLOR_ALIAS = "brand";
 
-export const ROOT_COLLECTION = {
+export const ROOT_COLLECTION: Collection = {
```
We shouldn't do it like that. The constant is used in the app for real and setting values `can_write: false` affects the runtime behavior. Furthermore, it's on purpose not a `Collection`, it's only a constant with a set of properties that the root collection has. 

If you need to pass it to somewhere in test where a `Collection` is expected, you need to use `createMockCollection(ROOT_COLLECTION)`

ranquild reviewed (changes requested):
`ROOT_COLLECTION` should not be changed

kamilmielnik committed:
Revert changes to ROOT_COLLECTION constant
- use createMockCollection

ranquild commented on enterprise/frontend/src/metabase-enterprise/sandboxes/components/EditSandboxingModal/EditSandboxingModal.unit.spec.tsx:
```diff
@@ -40,27 +50,30 @@ const TEST_CARD = createMockCard({
 });
 
-const setup = ({
+const TestComponent: typeof EditSandboxingModal = props => {
```
It's not clear why we're changing this. We should only preload data via the hooks in the test component if the component expects its to be loaded and doesn't handle loading the data himself. Is that's the case here? I've quickly went over `EditSandboxingModal` and couldn't find that it expects them to be preloaded.

If the component handles the data loading (e.g. using the entity loader / hook) it should be fine just to use `setupCollectionsEndpoints`

kamilmielnik replied:
`EditSandboxingModal` uses `QuestionPicker` which uses `ItemPicker` which triggers `Collections.selectors.getExpandedCollectionsById`, which I modified, which now uses `Collections.selectors.getList`, which uses `entity.selectors.getObject` which  requires `collections_list` to be present in the redux store.

kamilmielnik replied:
Without this rendering will crash because of data missing from redux store.

kamilmielnik replied:
> Is that's the case here? 

Yes, I believe so.

ranquild replied:
@kamilmielnik if the above is true, `ItemPicker` should be fixed and it should call `Collections.loadList` or new `useCollectionListQuery` - there shouldn't be a dependency like this. Can you fix this?

kamilmielnik replied:
Ok, now I get it.

Turns out there was a bug and the error message was misleading. I reverted these changes.

Good catch, thanks :+1: 

7c030c1fcd3fbb5a5ee03087e4ac66caf14857e1

ranquild reviewed (commented):
Unclear test changes

kamilmielnik committed:
Fix typing

kamilmielnik committed:
Remove dead link to Source Sans Pro font
- it caused markdown check to fail during CI, this fixes it
- it seems that the font has been removed from Google Fonts
- users can still choose this font in Metabase settings
- removal confirmation: https://metaboat.slack.com/archives/C5XHN8GLW/p1686320576992299?thread_ts=1686319384.637909&cid=C5XHN8GLW

kamilmielnik committed:
Merge branch 'master' into 31294-alphabetical-collections-in-move-modal

ranquild reviewed (commented):
Commented inline

kamilmielnik committed:
Fix getExpandedCollectionsById selector

kamilmielnik committed:
Revert adding useCollectionListQuery in unit tests

cdeweyx reviewed (approved):
Working as expected, nice job

kamilmielnik committed:
Merge branch 'master' into 31294-alphabetical-collections-in-move-modal
